astropy/coordinates/sky_coordinate.py

```diff
@@ -41,6 +42,37 @@ def FRAME_ATTR_NAMES_SET():
     return out
 
 
+class SkyCoordDataInfo(BaseInfo):
+    attrs_from_parent = set(['unit'])
+
+    @staticmethod
+    def default_format(val):
```

Comment:
There is some messiness due to a design decision at the coordinate Frame level which makes the user interface more intuitive at the expense of code nastiness.  Basically you can say the Frame has a SphericalDistance representation, but if you don't provide a distance at the time of creation then the actual data representation is UnitSphericalDistance.  In this special case there are shenanigans to make the repr correct.

But the other important bit is that the Frame (or SkyCoord) representation is mutable, but the actual data values (`skycoord.frame._data`) are never touched.  So a different representation of the coord data are generated on the fly (and cached) as needed.  That's what's happening in `_repr_data`.  This code is basically taken from the Frame `__repr__`.

Your idea about making `default_format` be a property returning a string does have merit in showing the user something useful that they can build on.  There is a detail in your implementation that would need fixing, which is that the `repr_data` components are things like `lon`, `lat` while what would be needed in the format statement are the frame names like `ra`, `dec`.  There is just a translation.

The bigger issue is about performance.  As you probably know iterating over a SkyCoord one element at a time is super slow because it creates a full-blown SkyCoord each time.  I have an idea to allow a format function to take an additional argument which is the index.  If there, it was assume the prior arg is the entire column.  So in coordinates it is *much* faster to do `c.ra[i]` than `c[i].ra`.  So using a function here instead of the format string would allow this optimization.  But I want to save that for a future PR.  Maybe speed is not the most crucial thing, but it's certainly possible someone could be writing a million-row table with a coordinate in `io.ascii`.